File: sdks/python/src/fireweave/domain/decision.py
"""Canonical evaluation decision (spec/decision.schema.json)."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Optional

from .errors import ErrorKind
from .types import FlagMetadata

__all__ = ["Decision", "Reason"]


class Reason:
    """Canonical reason strings (spec/decision.schema.json)."""

    TARGETING_MATCH = "TARGETING_MATCH"
    SPLIT = "SPLIT"
    DISABLED = "DISABLED"
    DEFAULT = "DEFAULT"
    STALE = "STALE"
    CACHED = "CACHED"
    STATIC = "STATIC"
    ERROR = "ERROR"


@dataclass(frozen=True)
class Decision:
    """Result of a flag evaluation. Evaluation APIs return this, never raise
    (spec/control-points.md "Return discipline — never throw into a read
    path")."""

    value: Any
    variant: Optional[str] = None
    reason: str = Reason.DEFAULT
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    error_kind: Optional[ErrorKind] = None
    flag_metadata: FlagMetadata = field(default_factory=dict)

    @property
    def is_error(self) -> bool:
        return self.reason == Reason.ERROR

File: sdks/swift/Sources/Fireweave/Domain/Decision.swift
/// Canonical evaluation decision (`spec/decision.schema.json`).

/// Canonical reason strings (`spec/decision.schema.json`).
public enum DecisionReason: String, Sendable, Equatable {
  case targetingMatch = "TARGETING_MATCH"
  case split = "SPLIT"
  case disabled = "DISABLED"
  case defaultReason = "DEFAULT"
  case stale = "STALE"
  case cached = "CACHED"
  case staticReason = "STATIC"
  case error = "ERROR"
}

/// Result of a flag evaluation. Evaluation APIs return this, never throw
/// (`spec/control-points.md` "Return discipline — never throw into a read
/// path").
public struct Decision: Sendable, Equatable {
  public var value: JSONValue
  public var variant: String?
  public var reason: DecisionReason
  public var errorCode: String?
  public var errorMessage: String?
  public var errorKind: ErrorKind?
  public var flagMetadata: FlagMetadata

  public init(
    value: JSONValue,
    variant: String? = nil,
    reason: DecisionReason,
    errorCode: String? = nil,
    errorMessage: String? = nil,
    errorKind: ErrorKind? = nil,
    flagMetadata: FlagMetadata = [:]
  ) {
    self.value = value
    self.variant = variant
    self.reason = reason
    self.errorCode = errorCode
    self.errorMessage = errorMessage
    self.errorKind = errorKind
    self.flagMetadata = flagMetadata
  }

  public var isError: Bool { reason == .error }
}
